### ISSUE TITLE
Add ability to customize build tags from groovy pipelines

### ISSUE THREAD

**Issue by gzussa** [jenkinsci/datadog-plugin#22]
Migrated from https://github.com/DataDog/jenkins-datadog-plugin/issues/115

## Description

I need to change a tag value based on what triggered the build (i.e. what's the build cause). After a lot of tinkering, I'm doing something like this:
```groovy
def cause = 'something' // actually comes from currentBuild.rawBuild.getCause...
properties([
  [
    $class: 'DatadogJobProperty',
    tagProperties: "cause=${cause}"
  ]
])
```
And this sort of works, but isn't really reliable. It causes issues when multiple jobs run concurrently. The [**current** `DatadogJobProperty` value is used](https://github.com/DataDog/jenkins-datadog-plugin/blob/1c3744329ec118f2f62a8ec4b158cb493380e266/src/main/java/org/datadog/jenkins/plugins/datadog/DatadogUtilities.java#L237-L238) whenever a build finishes. That means that if job A starts with cause "A" and job B starts with cause "B" before A finishes running, then both jobs will use the last `DatadogJobProperty` which has "B" configured as the cause.

I also tried the following, but this fails to set the value altogether. The tag ends up as `cause=_cause`.
```groovy
env.CAUSE = 'something' // actually comes from currentBuild.rawBuild.getCause...
properties([
  [
    $class: 'DatadogJobProperty',
    tagProperties: 'cause=${CAUSE}'
  ]
])
```

**Comment by gzussa** [comment]
The most relevant PR to look at is https://github.com/DataDog/jenkins-datadog-plugin/pull/164/ (from the old repo).
The second related issue is https://github.com/DataDog/jenkins-datadog-plugin/issues/152

**Comment by provcristianmaluenda** [comment]
I just was thinking of a similar requirement. It would be great if we can use env vars in the "Global job tags".
Something like `(.*?)/(.*?)/(.*?),a:$,b:$2,c:$3,team:$TEAM_NAME` 🤔

**Comment by stpierre** [comment]
The hack @gzussa describes to set tags dynamically in the original post also no longer works in v1.0.x of the plugin, so even the little ability we had to do this is gone (unless we stick to 0.7.1).

**Comment by gzussa** [comment]
In order to fix this properly, i would love to better understand user scenario (note that i didn't created this issue, i just migrated it from our old repo to this new one).

Would anyone have a clear and simple scenario to share? Ideally something that we could use as a test scenario (groovy pipeline and/or jenkinsFile). The hack mentioned above wasn't very clean and ideally, i would like something more standard and generic.

@provcristianmaluenda that seems like a different feature request. Do you mind to open a distinct issue for it?

**Comment by provcristianmaluenda** [comment]
I created this [feature request](https://github.com/jenkinsci/datadog-plugin/issues/54).

Thanks @gzussa

**Comment by deiwin** [comment]
For me (author of the original issue) the use case was and still is the following.

We have a shared library that allows executing specific types of builds based on GitHub comments. More specifically it supports deploying a service when a user writes `!deploy` in a PR comment. To gather metrics about deploys we differentiate between regular builds and deploys with a `is_deploy` tag in Datadog. The problem is that the tag can currently get an invalid value if deploys and regular PR builds run concurrently. To avoid that, we abort all other builds whenever a deploy starts. The actual code for this can be seen [here](https://github.com/salemove/pipeline-lib/blob/a8a464c19b0618169f05f447b2c8eb49da8b54ef/vars/deployer.groovy#L30-L57).

For a simple scenario, however, something like this should do (haven't tested):
```
def isCommentTriggeredBuild = !!currentBuild.rawBuild.getCause(org.jenkinsci.plugins.pipeline.github.trigger.IssueCommentCause)
properties(
  pipelineTriggers([issueCommentTrigger('!trigger-build')]),
  [
    $class: 'DatadogJobProperty',
    enableProperty: true,
    tagProperties: "is_comment_triggered_build=${isCommentTriggeredBuild}"
  ]
)
```

**Comment by deiwin** [comment]
Oh, and the triggering requires https://github.com/jenkinsci/pipeline-github-plugin

**Comment by gzussa** [comment]
Ok, we will have to evaluate what the best approach is here.

**Comment by s2oBCN** [comment]
Hi! How is this PR? We need this feature too :)

**Comment by s2oBCN** [comment]
Well, in fact, we see that this PR  https://github.com/DataDog/jenkins-datadog-plugin/pull/164 seems ok for us 
Is this possible to merge?
Thank you!

**Comment by FlorianVeaux** [comment]
Hey all here, I spent some time playing around with some of the provided use cases and I was able to get a better understanding of the situation.

By setting `tagProperties: "key=${VALUE}"` (note the double quotes), Jenkins will resolve the environment variables on each build and will update `DatadogJobProperty`. This is **not** how `DatadogJobProperty` is supposed to work, its configuration is supposed to be static between builds.

For dynamic properties (like tagging), the plugin **does support** resolving environment variables. But to do so, you need to set `tagProperties: 'tag_key=${VALUE}'` (note the single quotes), so that the plugin can try to resolve environment variables on each build and have a dynamic value.

Yet, you've clearly noted that the plugin is not able to resolve variables (and replace `${XYZ}` with `_xyz`) in case of **pipeline** builds, so what's going on?
It appears that the environment where jenkins agents run the pipeline scripts and the environment where the pipeline (and the datadog plugin) run are different. So from the standpoint of the datadog plugin, there is no access to those env vars.
Some links pointing to that:
https://stackoverflow.com/a/60781824
https://stackoverflow.com/q/52704473

As pointed out by @s2oBCN the PR DataDog/jenkins-datadog-plugin#164 seems to go in the right direction, but as far as I can tell it does't seem to solve the problem as env vars resolution still takes place in the wrong scope. I think the way forward here is to add a DatadogBuildStep that will take care of resolving the env vars **during the run**.

**Comment by FlorianVeaux** [comment]
Hey there,
We've just released v2.0.0 of the Datadog plugin for Jenkins which provides better support for pipelines. It also adds a `datadog` step that can be used from within pipelines definition to set custom tags like so

- Declarative
```groovy
def cause = 'something';
pipeline {
    agent any
    options {
        datadog(collectLogs: true, tags: ["cause:${cause}"])
    }
    stages {
        stage('Example') {
            steps {
                echo "Hello world."
                sh 'sleep 15'
            }
        }
    }
}
```

- Scripted
```groovy
def cause = 'something'

datadog(collectLogs: true, tags: ["cause:${cause}"]){
  node {
    stage('Example') {
      echo "Hello world."
      sh 'sleep 15'
    }
  }
}
```
